File: basic_dsa_three_hundred_eighty_two.java
/**
 * Basic DSA Problems - Linked List Random Node (LeetCode 382)
 * getRandom() returns random node value with equal probability
 */
public class basic_dsa_three_hundred_eighty_two {

    // Problem: getRandom() - each node has equal probability.

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    static class Solution {
        ListNode head;
        java.util.Random rnd = new java.util.Random();

        Solution(ListNode head) { this.head = head; }

        /**
         * Reservoir sampling: for node i, replace with prob 1/i.
         * Time O(n), Space O(1)
         */
        public int getRandom() {
            int out = head.val;
            ListNode cur = head.next;
            for (int i = 2; cur != null; i++, cur = cur.next)
                if (rnd.nextInt(i) == 0) out = cur.val;
            return out;
        }
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Linked List Random Node");
        System.out.println("=".repeat(60));

        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        Solution sol = new Solution(head);
        System.out.println("List [1,2,3], getRandom x5: " + sol.getRandom() + " " + sol.getRandom()
                + " " + sol.getRandom() + " " + sol.getRandom() + " " + sol.getRandom());
    }
}

File: basic_dsa_three_hundred_eighty_two.cpp
/**
 * Basic DSA - Count Binary Substrings (LeetCode 696)
 * Count substrings with equal number of consecutive 0s and 1s.
 */
#include <algorithm>
#include <iostream>
#include <string>

int countBinarySubstrings(const std::string& s) {
    int prev = 0;
    int cur = 1;
    int ans = 0;
    for (size_t i = 1; i < s.size(); ++i) {
        if (s[i] == s[i - 1]) {
            ++cur;
        } else {
            ans += std::min(prev, cur);
            prev = cur;
            cur = 1;
        }
    }
    return ans + std::min(prev, cur);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Count Binary Substrings\n";
    std::cout << std::string(60, '=') << "\n";

    std::cout << "\"00110011\" -> " << countBinarySubstrings("00110011") << "\n";
    std::cout << "\"10101\" -> " << countBinarySubstrings("10101") << "\n";
}
